Fix unsetting cgroup restrictions

Werkov opened the pull request:
This series aims to tackle #14149 and #14917. The issues are related but approach is different because they happen with v1 and v2 hierarchies respectively. Rough idea is to use cgroup realization queue to perform changes on modified unit's ancestors (disabling must happen bottom-up). See individual commit messages for details.

The current implementation likely introduces performance slowdown and the series (after review) should be accompanied with related changes.

[Scripts](https://gist.github.com/Werkov/001d1c6ce28a74c35ac9c246d592899f) checking the behavior, they aren't full independent tests.

Werkov commented:
Force push to fix tests and rebase from v245 to master.

cdown commented on src/shared/cgroup-setup.c:
```diff
@@ -674,51 +672,38 @@ int cg_attach_many_everywhere(CGroupMask supported, const char *path, Set* pids,
+                        continue;
 
-                (void) cg_migrate_recursive_fallback(SYSTEMD_CGROUP_CONTROLLER, to, cgroup_controller_to_string(c), p, 0);
+                to = to_callback(bit, userdata);
```
No need for this any more, you can just pass it to `cg_migrate_recursive_fallback`.

Werkov replied:
I'd like to get rid of the callback too. `*_fallback` would work, however, it seems not to be super effective, PIDs are enumerated before a non-existence is checked. Hence I sticked to the path based on realized path.

cdown commented on src/core/cgroup.c:
```diff
@@ -1860,20 +1867,30 @@ static int unit_create_cgroup(
+                r = cg_trim_controllers(u->manager->cgroup_supported, ~target_mask, u->cgroup_path, !is_root_slice);
                 if (r < 0)
-                        log_unit_warning_errno(u, r, "Failed to migrate cgroup from to %s, ignoring: %m", u->cgroup_path);
+                        log_unit_warning_errno(u, r, "Failed to delete cgroup %s, ignoring: %m", u->cgroup_path);
```
Is this even possible? It looks like `cg_trim_controllers` always returns 0.

Werkov replied:
Good question. I've modified the code so that v1 controller hierarchies migration and removal errors are caught and reported to be on the safe side.

cdown commented on src/core/cgroup.c:
```diff
@@ -1629,7 +1629,10 @@ const char *unit_get_realized_cgroup_path(Unit *u, CGroupMask mask) {
+        /* If not realized at all, migrate to root .
+         * It may happen if we're upgrading from older version that didn't clean up.
+         */
+        return unit_get_realized_cgroup_path(userdata, mask) ?: "";
```
Hmm, I'm not sure how I feel about this one, since it's hard to specify that as an invariant which we can actually check at runtime. I'm not sure how risky it is, I need to look more at how we end up here.

cdown replied:
Is this a practical situation you actually ran into, or just a theoretical concern?

Werkov replied:
The blkio root cgroup wasn't realized when I tested with v245 on TW. After I reexeced my modified version I `to_callback` returned NULL and that wasn't well accepted (assert crash).

cdown commented:
I've now been through each commit individually and don't have any additional comments, this looks really good. Thanks also for providing clear changelogs, it makes it a lot easier to go through.

Other than the above comments, the only other thing I'd like to see is some tests which validate this new functionality added to the CI.

Adding needs-rework just for the comments above. Thanks!

poettering commented on src/core/unit.c:
```diff
@@ -653,6 +645,12 @@ void unit_free(Unit *u) {
+         * properly. Do this after we detach the unit from slice tree in order to eliminate its effect on
+         * controller masks. */
+        if (u->type != _UNIT_TYPE_INVALID && UNIT_HAS_CGROUP_CONTEXT(u))
+                unit_add_siblings_to_cgroup_realize_queue(u);
```
the comment is misleading: the slice is still semi-attached to the unit, the `unit_ref_unset(&u->slice);` call comes a few lines further down only.

I'd prefer if we'd change this around a bit, so that we don't call unit_add_siblings_to_cgroup_realize_queue() on a semi-destroyed unit (which `u` is at this point.)

maybe we should split unit_add_siblings_to_cgroup_realize_queue() up, and have

```c
void slice_add_family_to_cgroup_realize_queue(Slice *slice) {
        /* We define "family" as the slice's immediate children, the slice itself, and all parents all the way up the tree, as well as all siblings of these. */
        …
}
void unit_add_siblings_and_parents_to_cgroup_realize_queue(u) {
        slice_add_family_to_cgroup_realize_queue(u->slice);
}
```

and then call slice_add_family_to_cgroup_realize_queue(…) here on the slice object, after we *fully* detached our unit from it. That way things should be more robust, as we won't run into trouble with "half-destroyed" objects ever.

poettering replied:
if you do it like that then you can drop the special "self" handling you added to  unit_add_siblings_and_parents_to_cgroup_realize_queue(), right?

Werkov replied:
Naming the set we need to operate on is a good idea. I introduced family only (no siblings function) and explained in the comments its effect and purpose.

I've left the call-site in `unit_free` the same. There is (and was) a "race" between slice GC queue processing and cgroup realization queue. That should be OK:
- if GC goes before CG, then the unit is removed from CG queue (but it should have been `unit_prune_cgroup`'d by then) and ancestors are still CG queued,
- if CG goes before GC, then it's like the normal case (i.e. when no GC happened).

poettering commented on src/core/cgroup.c:
```diff
@@ -2314,10 +2304,12 @@ unsigned manager_dispatch_cgroup_realize_queue(Manager *m) {
-static void unit_add_siblings_to_cgroup_realize_queue(Unit *u) {
+void unit_add_siblings_to_cgroup_realize_queue(Unit *u) {
         Unit *slice;
+        Unit *self = u;
```
doesn't matter much, but this looks backwards. i.e. instead of keeping a copy of the original pointer we shouldn't even modify the original pointer but a copy of it. i.e. intrdouce `Unit *p = u;` here, and then keep updating `p` instead of `u` when iterating through the tree.

Werkov replied:
Thanks, avoided that with the family rework (but used that in the intermediate commits).

poettering commented on src/core/cgroup.c:
```diff
@@ -2332,10 +2324,15 @@ static void unit_add_siblings_to_cgroup_realize_queue(Unit *u) {
                 Unit *m;
                 void *v;
 
+                slice->cgroup_members_mask_valid = false;
```
this line deserves a comment

poettering replied:
or instead the big comment above should explicitly say what your comit msg say: this also implicitly invalidates all members masks

Werkov replied:
This made me review my original removal of mask invalidation from `unit_load`. I realized that wouldn't be correct. I reorganized the commits a bit and added the explanation in the message of "cgroup: Implicit unit_invalidate_cgroup_members_masks".

poettering commented on src/shared/cgroup-setup.c:
```diff
@@ -731,7 +716,15 @@ int cg_trim_everywhere(CGroupMask supported, const char *path, bool delete_root)
+        return cg_trim_controllers(supported, _CGROUP_MASK_ALL, path, delete_root);
+}
+
+int cg_trim_controllers(CGroupMask supported, CGroupMask mask, const char *path, bool delete_root) {
```
hmm, this call is now cgroupsv1 specific, unlike most other calls here, and there's not even an assert() to check it's not called on the wrong environment... or clarified in the name or so

Werkov replied:
Distinguished the functions by their name.

poettering commented on src/shared/cgroup-setup.c:
```diff
@@ -674,51 +672,38 @@ int cg_attach_many_everywhere(CGroupMask supported, const char *path, Set* pids,
 }
 
-int cg_migrate_everywhere(CGroupMask supported, const char *from, const char *to, cg_migrate_callback_t to_callback, void *userdata) {
+int cg_migrate_controllers(CGroupMask supported, CGroupMask mask, const char *from, cg_migrate_callback_t to_callback, void *userdata) {
```
hmm, so this call becomes cgroupvs1-only this way. and there's no indication in name or assert or so.

Werkov replied:
Distinguished the functions by their name.

poettering commented on src/shared/cgroup-setup.c:
```diff
@@ -674,51 +672,38 @@ int cg_attach_many_everywhere(CGroupMask supported, const char *path, Set* pids,
-                return q;
-        if (q > 0)
-                return r;
+        assert(to_callback);
```
i think this function should have a comment explaining what it does, i.e. that it only and exclusively migrates the processes in the cgroupsv1 controller hierarchies, i.e. not in the name=systemd hierarchy.

Werkov replied:
Should be clear from the name now.

Werkov commented on src/core/cgroup.c:
```diff
@@ -2112,7 +2112,7 @@ void unit_add_to_cgroup_realize_queue(Unit *u) {
                 return;
 
-        LIST_PREPEND(cgroup_realize_queue, u->manager->cgroup_realize_queue, u);
+        LIST_APPEND(cgroup_realize_queue, u->manager->cgroup_realize_queue, u);
```
FTR, it's not the only misnomer. Other usages likely don't care about order (didn't check though).

Werkov commented on src/core/cgroup.c:
```diff
@@ -2360,6 +2341,9 @@ static void unit_add_siblings_to_cgroup_realize_queue(Unit *u) {
         }
+
+        /* Root slice comes last */
+        unit_add_to_cgroup_realize_queue(u);
```
family FTW

Werkov commented:
Thanks for your review and good ideas. Hopefully, I addressed all your comments in the force pushed branch now (the second force push was just commit message update).

Werkov commented:
The tag `reviewed/needs-rework` is stale, i.e. this is ready for another review.

(With v246-rc1 out, I assume this will only get into the v247, I'd like just to update the issue tags so that it isn't forgotten. Thanks.)

poettering commented on src/core/cgroup.c:
```diff
@@ -1629,7 +1629,10 @@ const char *unit_get_realized_cgroup_path(Unit *u, CGroupMask mask) {
+        /* If not realized at all, migrate to root .
+         * It may happen if we're upgrading from older version that didn't clean up.
+         */
+        return unit_get_realized_cgroup_path(userdata, mask) ?: "";
```
strempty()

poettering commented:
any chance you can make the minor strempty() change, rebase and repush. will merge then as soon as CI passes
